src/snowflake/snowpark/_internal/udf_utils.py

```diff
@@ -87,39 +93,140 @@ def is_local_python_file(file_path: str) -> bool:
 def get_types_from_type_hints(
     func: Union[Callable, Tuple[str, str]],
     object_type: TempObjectType,
+    output_schema: Optional[List[str]] = None,
 ) -> Tuple[DataType, List[DataType]]:
     if isinstance(func, Callable):
         # For Python 3.10+, the result values of get_type_hints()
         # will become strings, which we have to change the implementation
         # here at that time. https://www.python.org/dev/peps/pep-0563/
+        func_name = func.__name__
         try:
-            python_types_dict = get_type_hints(
-                getattr(func, TABLE_FUNCTION_PROCESS_METHOD, func)
-            )
+            if object_type == TempObjectType.TABLE_FUNCTION:
+                if not (
+                    hasattr(func, TABLE_FUNCTION_END_PARTITION_METHOD)
+                    or hasattr(func, TABLE_FUNCTION_PROCESS_METHOD)
+                ):
+                    raise AttributeError(
+                        f"Neither `{TABLE_FUNCTION_PROCESS_METHOD}` or `{TABLE_FUNCTION_END_PARTITION_METHOD}` is defined for class {func}"
+                    )
+                python_types_dict = {}
+                if hasattr(func, TABLE_FUNCTION_END_PARTITION_METHOD):
```

Comment:
This means we will now prefer `TABLE_FUNCTION_END_PARTITION_METHOD`'s result, whereas previously we use `TABLE_FUNCTION_PROCESS_METHOD`. Is this a concern?

Reply:
Right 🤔  this is a problem for `input types` since that could be different for `process` and `end_partition`. Let me add a check.